### PR TITLE
IDEMPIERE-5859 - Support Attribute Set Instance for Reference=>Multi-Chosen fieldtypes

### PR DESCRIPTION
**Pull request by PeterTakacs300**
JIRA ticket: https://idempiere.atlassian.net/browse/IDEMPIERE-5859

# Pull Request Checklist

- [x] My code follows the [code guidelines](https://wiki.idempiere.org/en/Contributing_to_Trunk) of this project
- [x] My code follows the best practices of this project
- [x] I have performed a self-review of my own code
- [x] My code is easy to understand and review. 
- [x] In hard-to-understand areas, I have commented my code.
- [x] My changes generate no new warnings
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added unit tests that prove my fix is effective or that my feature works
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings
- [ ] I have made corresponding changes to the documentation as follows:
  - [ ] New feature (non-breaking change which adds functionality): I have created the New Feature page in the project wiki explaining the functionality and how to use it. If relevant, I have committed sample data to the core seed to have usable examples in GardenWorld.
  - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected): I have documented the change in a clear way that everyone in the community can understand the impact of the change.
  - [ ] Improvement (improves and existing functionality): This documentation is needed if the improvement changes the way the user interacts with the system or the outcome of a process/task changes. If it is just, for instance, a performance improvement, documentation might not be needed. 
- [ ] The changed/added documentation is in the project wiki (not privately-hosted pdf files or links pointing to a company website) and is complete and self-explanatory.

### CODE REVIEW

**Comment by PeterTakacs300** [comment]
Currently the description is set as the IDs of the multi-selection values (this comes from the default behaviour of this field type).

I suggest to add a new field to M_AttributeInstance that will store the display value of multi-selection. I want to add a new field because of the way the Description is parsed in MAttributeSetInstance here: [MAttributeSetInstance.java:235](https://github.com/idempiere/idempiere/blob/master/org.adempiere.base/src/org/compiere/model/MAttributeSetInstance.java#L235)

Could not figure out any better way to get the display value into the MAttributeSetInstance.setDescription() method than to store it in DB, but I wanted to discuss it before implementing it. Any ideas?

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/WPAttributeDialog.java [hunk 1]
```diff
@@ -695,21 +699,40 @@ public GridField getDateGridField(MAttribute attribute)
...
+	public GridField getListTypeGridField(MAttribute attribute)
+	{
+	    return getGridFieldForDisplayType(attribute, DisplayType.TableDir);
 	} // getListTypeGridField
 
+	/**
+	 * @param attribute
+	 * @return GridField for DisplayType.ChosenMultipleSelectionTable
+	 */
+	public GridField getMultiSelectionListTypeGridField(MAttribute attribute)
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/WPAttributeDialog.java:731]: should be private or protected

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/WPAttributeDialog.java [hunk 2]
```diff
@@ -695,21 +699,40 @@ public GridField getDateGridField(MAttribute attribute)
...
 		vo.lookupInfo.IsValidated = false;
 
 		return createGridField(attribute, vo);
+	} // getGridFieldForDisplayType
+
+	/**
+	 * @param attribute
+	 * @return GridField for DisplayType.TableDir
+	 */
+	public GridField getListTypeGridField(MAttribute attribute)
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/WPAttributeDialog.java:722]: should be private or protected

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/WPAttributeDialog.java [hunk 3]
```diff
@@ -695,21 +699,40 @@ public GridField getDateGridField(MAttribute attribute)
 
 	/**
 	 * @param attribute
-	 * @return GridField for DisplayType.TableDir
+	 * @param displayType
+	 * @return GridField for given displayType
 	 */
-	public GridField getListTypeGridField(MAttribute attribute)
+	public GridField getGridFieldForDisplayType(MAttribute attribute, int displayType)
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/WPAttributeDialog.java:705]: should be private or protected

File: org.adempiere.base/src/org/compiere/model/X_M_Attribute.java [hunk 4]
```diff
@@ -207,6 +207,8 @@ public int getAD_Val_Rule_ID()
 
 	/** AttributeValueType AD_Reference_ID=326 */
 	public static final int ATTRIBUTEVALUETYPE_AD_Reference_ID=326;
+	/** Chosen Multiple Selection List = CL */
+	public static final String ATTRIBUTEVALUETYPE_ChosenMultipleSelectionList = "C";
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/X_M_Attribute.java:211]: It seems you are missing the saveSelection changes for ATTRIBUTEVALUETYPE_ChosenMultipleSelectionList

**Comment by hengsin** [comment]
hi @PeterTakacs300 , https://github.com/idempiere/idempiere/blob/master/org.adempiere.base/src/org/compiere/model/MAttributeSetInstance.java#L235 shouldn't be a problem. It is now because you doesn't change saveSelection for ATTRIBUTEVALUETYPE_ChosenMultipleSelectionList and create M_Attributeinstance record for it.

File: org.adempiere.base/src/org/compiere/model/MAttributeInstance.java [hunk 5]
```diff
@@ -261,6 +277,16 @@ else if (displayType == DisplayType.Time)
...
 	} // setValueDate
+	
+	/**
+	 * Set valueMultipleSelection
+	 * @param value - comma separated IDs
+	 * @param displayValue - comma separated display values
+	 */
+	public void setValueMultipleSelection(String value, String displayValue) {
+		super.setValueMultipleSelection(value);
+		setValue(displayValue);
```
[the remark this answers is not in the document]
Comment (PeterTakacs300) [reply org.adempiere.base/src/org/compiere/model/MAttributeInstance.java:288]: increased m_attributeinstance.value from 40 to 255, and m_attributesetinstance.description from 255 to 1023

File: org.adempiere.base/src/org/compiere/model/MAttributeInstance.java [hunk 6]
```diff
@@ -261,6 +277,16 @@ else if (displayType == DisplayType.Time)
 
 		setValue(sdf.format(valueDate));
 	} // setValueDate
+	
+	/**
+	 * Set valueMultipleSelection
+	 * @param value - comma separated IDs
+	 * @param displayValue - comma separated display values
+	 */
+	public void setValueMultipleSelection(String value, String displayValue) {
```
[the remark this answers is not in the document]
Comment (PeterTakacs300) [reply org.adempiere.base/src/org/compiere/model/MAttributeInstance.java]: renamed to `setMultiSelectValueAndDisplay`

**Review by hengsin — APPROVED** [review]
tested

**Comment by PeterTakacs300** [comment]
Hi @hengsin, sadly I've forgotten to turn on the migration script logging at some point of the setup. The attribute value tab of the Attribute window must not be read-only if the Attribute Value Type Chosen Multiple Selection List is selected. 

Missing setup committed here: https://github.com/idempiere/idempiere/pull/2251

**Pull request merged**